### DISCUSSION TITLE
data.json file is getting big and causing opa latency

### DISCUSSION THREAD

**Discussion by itayhac** [open-policy-agent/community#420] [OPA and Rego]
i got a bundle with the following structure:
<img width="518" alt="image" src="https://github.com/open-policy-agent/community/assets/45124094/199b9f0a-2154-47f9-b0f9-50a027d861d6">

the data.json's size is usually ~2-5KB and total evaluation time is <10ms.
but recently i faced an issue with data.json's size getting to ~1MB , which makes opa very slow, which in turn makes opa's total evaluation time ~ 300ms.

after running opa eval with profiler it looks like the import of the data.json file to the logic.rego takes most of the time (the 300ms).
need to say that the data.json does not change during a revision of bundle.

im using the following command for debugging:
`opa eval -d  . -i input.json  "data.logic.access" --profile --format=pretty --count=1`

i can see using the profiler that almost of all the time is consumed by the `constraints = object.get(index, "constraints", {})` line of code. meaning that the dynamic import of the data set is slowing us down.

the question is as followed:  

how can i improve the latency of opa component in our system? 
first thing that came to my mind is to "hard code" the data.json in our rego file, instead of importing it.

<img width="298" alt="image" src="https://github.com/open-policy-agent/community/assets/45124094/4a36c2af-aa17-46ec-a216-55b95adb25e8">

```
package logic.access

import data.my_data.index

access[decision]{
  constraints = object.get(index, "constraints", {})
  decision := access_buisness_logic(constraints)
}

access_buisness_logic(constraints) = ret {
    # some logic here #
    ret := {
        "result": "allowed",
    }
}
```

**Comment by charlieegan3** [comment]
Hey @itayhac, 1MB isn't a very large dataset so my guess is that the policy is using some inefficient logic. Have you tried reviewing the recommendations in https://www.openpolicyagent.org/docs/latest/policy-performance/ ?

**Reply by itayhac** [reply]
when running opa eval with --profile tag i can see that  this line takes almost all of the time.
`constraints = object.get(index, "constraints", {})
`
i can clearly see even if i remove the logic function, the time remains the same.
so my question is : if the data set import takes a lot of time, can i do something to optimize it? can i use caching? so even if the data set import if "heavy" at least it would happen only once?

**Comment by anderseknert** [comment]
I'm not sure I follow here — and your screenshot is cut off before any metrics are shown. 

Are you using `opa eval`? You'd normally do that either in development or in CI/CD, and 300 ms wouldn't likely be considered much in that context. If you could expand some on your use case, and the commands you execute, that'd be helpful 🙂

**Reply by itayhac** [reply]
im using both opa eval for debugging and opa run for production.
both give me same result:
constraints = object.get(index, "constraints", {})
this line of code takes 200+ ms.
also, there's  no data set caching, so every evaluation is taking a lot of time.
any idea how to optimize it? does rego support data set caching?

**Reply by anderseknert** [reply]
The data is stored in memory when using `opa run` so there should be no need to re-evaluate that with each query. Perhaps `object.get` is doing some shenanigans here. Could you try and exclude that from the equation? Something like:

```rego
default constraints := {}

constraints := index.constraints
```

And see if it makes any difference?

**Reply by itayhac** [reply]
i figured a valid example would be the best way to proceed :)
here is a very simple rego code with a large data set.

**run the server:**
`opa run -s -l debug slow_bundle.tar.gz`

**trigger opa evaluation:**
`curl localhost:8181/v1/data/opa/access`

now you should see about ~150 ms latency. (im using `curl localhost:8181/metrics` in order to find out the latency, im looking at the http_request_duration_seconds_sum metric).

you can see that i have took your suggestion and declared the constraints variable as global..it still has the same problem.


now, if i "copy" the big dataset and paste it in the evaluator.rego file (hard coded) the latency drops to ~3ms .

what am i missing? does opa not support big data set? 

any help would be much appreciated!


attached is the slow_bundle.tar.gz file , as described above you should be able to run it via opa run
[slow_bundle.tar.gz](https://github.com/open-policy-agent/community/files/11533605/slow_bundle.tar.gz)

**Reply by anderseknert** [reply]
Thanks for the data, and the bundle to test!

Running with your bundle:
```shell
❯ opa run -s -l debug slow_bundle.tar.gz
```

And your query, with metrics reporting enabled:
```shell
❯ curl 'localhost:8181/v1/data/opa/access?metrics=true&pretty=true'
{
  "metrics": {
    "counter_server_query_cache_hit": 1,
    "timer_rego_external_resolve_ns": 374,
    "timer_rego_input_parse_ns": 375,
    "timer_rego_query_eval_ns": 108000,
    "timer_server_handler_ns": 146375
  },
  "result": [
    [
      "ALLOW"
    ]
  ]
}
```
So, in total a `timer_server_handler_ns` value of 146375, or 0.15 milliseconds.

**Reply by itayhac** [reply]
you are correct, apparently there's a difference between compressed/uncompressed file. 
but the problem remains the same with a bigger file.
here's a new bundle, you can see its bigger. i still get 200ms +.


run > `opa run -s -l debug slow_bundle.tar.gz`
query > `curl localhost:8181/v1/data/evaluator/access`


[slow_bundle.tar.gz](https://github.com/open-policy-agent/community/files/11543100/slow_bundle.tar.gz)

thanks a lot for your help!

**Reply by anderseknert** [reply]
The one evaluates in 1.9 milliseconds:

```
❯ curl 'localhost:8181/v1/data/opa/access?metrics=true&pretty=true'
{
  "metrics": {
    "counter_server_query_cache_hit": 1,
    "timer_rego_external_resolve_ns": 875,
    "timer_rego_input_parse_ns": 750,
    "timer_rego_query_eval_ns": 1871209,
    "timer_server_handler_ns": 1943208
  }
}
```
I think you might be looking at the wrong place. Can you run with the same query parameters as seen below? That way you'll get the evaluation time back in the response.

**Reply by itayhac** [reply]
ok now that i understand how to query opa and get metrics in the response:

i have:
1. bundle with imported dataset - takes 200ms 
ways to reproduce: 
run > `opa run -s -l debug slow_bundle.tar.gz`
query > `curl 'localhost:8181/v1/data/evaluator/access?metrics=true&pretty=true'` 

2. bundle with hard coded dataset - takes <1ms 
ways to reproduce: 
[fast_bundle.tar.gz](https://github.com/open-policy-agent/community/files/11544892/fast_bundle.tar.gz)
[slow_bundle.tar.gz](https://github.com/open-policy-agent/community/files/11544897/slow_bundle.tar.gz)

run > `opa run -s -l debug fast_bundle.tar.gz`
query > `curl 'localhost:8181/v1/data/evaluator/access?metrics=true&pretty=true'` 

thank you so much for your help!

**Reply by anderseknert** [reply]
Please attach the full output of running both of those commands. I've seen nothing close to 200 ms.

**Reply by itayhac** [reply]
slow command: 
<img width="891" alt="image" src="https://github.com/open-policy-agent/community/assets/45124094/a026dce2-4624-4468-b61c-fc8eac7a9040">

fast command:
<img width="959" alt="image" src="https://github.com/open-policy-agent/community/assets/45124094/cf1e3c3f-282d-4794-8dc6-e1a854d82e37">

**Reply by anderseknert** [reply]
Ah, I see it now, thanks! Will take a look tomorrow.
